### DISCUSSION TITLE
New Face Mesh model from Mediapipe

### DISCUSSION THREAD

**Discussion by ButzYung** [vladmandic/human#185] [General]
https://github.com/google/mediapipe/releases/tag/v0.8.8

At the first glance, it seems they have (finally) integrated the iris/blink detection into the face model. However, the model description seems to suggest that it's actually a new model using a new approach. More accurate lips and general 3D coordinates perhaps?

https://google.github.io/mediapipe/solutions/face_mesh.html#attention-mesh-model

**Comment by vladmandic** [comment]
Standard **FaceMesh** is 468 points, then its up to post processing to 
- cut eye boxes
- execute separate **Iris** model 
- recombine points for total of 478 points  

new **FaceMesh-With-Attention** is simmilar except it does that inside the model  
without post-processing and manual triggering of each sub-model:  
- It runs **FaceMesh**, then cuts and processes specific areas with specific parts of the model  
  It does so for **eye**, **iris**, and **lips** (and has no effect at all on the rest of the face) and result is  
  the same 478 points but with more preceise points for eyes, irises and lips  

Size wise, its double the size of normal FaceMesh, but still within reason  

But...MediaPipe releases only *TFLite* models and converting this is not going to be easy as multiple models are linked with a custom op `Landmarks2TransformMatrix` that doesn't exist in normal TensorFlow  
I've asked Pinto to take a look as he's an expert in TFLite models: <https://github.com/PINTO0309/PINTO_model_zoo/issues/143>

And I just noticed there are now two variations of **BlazeFace** detector full variation as it has been  
forked into *"dense"* and *"sparse"* variations  
That's mostly related to which ops are used and what is the performance impact depending on the backend  
(one is faster on GPU and another on WASM)  
I'll test it but but I believe impact of that would be minimal since detector boxes are mostly cached anyhow  

Btw, there is also new **FaceGeometry** model which runs as calculator that takes mesh and performs  
camera transforms into "real" 3D space  

Intention is more precise AR usage such as wearing fake glasses and what not  
(Currently no plans for integration of FaceGeometry into Human)

**Comment by vladmandic** [comment]
Just closing the loop...   

`facemesh-with-attention` uses several custom C functions up to the point that I don't even consider it a TFLite model as it requires a custom backend from `MediaPipe`  

Don't know why MediaPipe is going in that direction instead of contributing to TF, but as it is, it might take someone to reimplement the model from scratch (it is documented in sources) as an easier option than trying to convert it from their TFLite binary

**Comment by ButzYung** [comment]
Did some testings and it doesn't seem noticeably better than the current solution in terms of accuracy. A unified model has its advantage though, especially for the Mediapipe JS solutions like `Holistic`.

It will be interesting to see if the official TFJS side will do a conversion or not.

**Comment by vladmandic** [comment]
TFJS side stopped doing MediaPipe conversions last year, all of contributions since then are directly from MediaPipe team  
(see *Publisher* field on <https://tfhub.dev/s?q=mediapipe>)  
So its more of a question for them - and given they chose to use custom ops, I doubt it.
